src/ess/reduce/uncertainty.py

```diff
@@ -36,50 +38,99 @@
 
 @overload
 def broadcast_with_upper_bound_variances(
-    data: sc.Variable, sizes: Dict[str, int]
+    data: sc.Variable, prototype: sc.DataArray | sc.Variable
 ) -> sc.Variable:
     pass
 
 
 @overload
 def broadcast_with_upper_bound_variances(
-    data: sc.DataArray, sizes: Dict[str, int]
+    data: sc.DataArray, prototype: sc.DataArray | sc.Variable
 ) -> sc.DataArray:
     pass
 
 
 def broadcast_with_upper_bound_variances(
-    data: Union[sc.Variable, sc.DataArray], sizes: Dict[str, int]
+    data: Union[sc.Variable, sc.DataArray], prototype: sc.DataArray | sc.Variable
 ) -> Union[sc.Variable, sc.DataArray]:
-    if _no_variance_broadcast(data, sizes):
+    """
+    Compute an upper bound for the variances of the broadcasted data.
+
+    The variances of the broadcasted data are computed by scaling the variances of the
+    input data by the volume of the new subspace. The volume of the new subspace is
+    computed as the product of the sizes of the new dimensions.
+
+    Parameters
+    ----------
+    data:
+        The data to broadcast.
+    prototype:
+        Defines the new sizes (dims and shape). If present, masks are used to exclude
+        masked values from the variance computation.
+
+    Returns
+    -------
+    :
+        The data with the variances scaled by the volume of the new subspace.
+    """
+    if _no_variance_broadcast(data, prototype.sizes):
         return data
-    size = 1
+    sizes = {dim: size for dim, size in prototype.sizes.items() if dim not in data.dims}
+    sizes = prototype.sizes
+    mask = sc.scalar(False)
+    if isinstance(prototype, sc.DataArray):
+        if (irred := irreducible_mask(prototype, dim=sizes)) is not None:
+            for dim in data.dims:
+                if dim in irred.dims:
+                    irred = irred.all(dim)
+            mask = irred
+    size = (~mask).sum().value
     for dim, dim_size in sizes.items():
-        if dim not in data.dims:
+        if dim not in data.dims and dim not in mask.dims:
             size *= dim_size
     data = data.copy()
     data.variances *= size
-    return data.broadcast(sizes={**sizes, **data.sizes}).copy()
+    sizes = {**sizes, **data.sizes}
+    data = data.broadcast(sizes=sizes).copy()
+    if mask is not None:
+        # The masked values are not counted in the variance, so we set them to infinity.
+        data.variances[mask.broadcast(sizes=sizes).values] = np.inf
```

Comment:
What's the reasoning behind `np.inf`? I know the masked data should not be used, is `np.nan` a better choice? Inf maybe be annoying to handle in plots, fx.

Reply:
I went by my gut feeling "I am making some error bars smaller, so the others should be larger, to balance it out". Wouldn't scipp/plopp#3292 have solved the issue with plotting you mention?

That being said, if you think NaN actually makes more sense then I can change it.

Reply:
Actually, it makes no difference for the plotting, so you can leave it if you like. I though you would get vertical lines spanning the entire height of the figure for the error bars, but in fact nothing is plotted (both my MPL and plotly).
![Screenshot at 2024-06-21 10-51-06](https://github.com/scipp/essreduce/assets/39047984/23a3f686-9050-4a5f-bae1-151a0c1d4be9)

I don't know what is a better indicator that the data is invalid there: if it's NaN, does it appear more 'intentional' than inf which could be a result from an operation gone wrong?
In any case, I think it's ok either way.